### PR TITLE
handling and returning valid error code for context.Canceled and cont…

### PR DESCRIPTION
**Pull request by chetanvalagit**
…ext.DeadlineExceeded error

*Issue #, if available: #293

*Description of changes:*
Proper error code return in case of context.Canceled (Error code will be returned twirp.Canceled) and context.DeadlineExceeded (Error code will be returned twirp. DeadlineExceeded), instead of 5xx 'Unexpected EOF'

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: internal/twirptest/proto/proto_test.go [hunk 1]
```diff
@@ -13,8 +13,129 @@
...
+	"time"
+
+	twirp "github.com/twitchtv/twirp"
+)
 
 func TestCompilation(t *testing.T) {
 	// Test passes if this package compiles
 }
+
+type errReader int
```
Comment (marioizquierdo) [inline internal/twirptest/proto/proto_test.go:33]: Adding the test in `/proto/proto_test.go`,  feels a little arbitrary, because this is not related to testing an empty proto message. What about moving the test to `service_test.go`? all other arbitrary tests are there

File: internal/twirptest/proto/proto_test.go [hunk 2]
```diff
@@ -13,8 +13,129 @@
...
+	ctx, _ = context.WithTimeout(ctx, 1*time.Millisecond)
+
+	// Make a request, that will endpoint method
+	req, _ := http.NewRequest(http.MethodPost, "http://testing:8080/twirp/twirp.internal.twirptest.proto.Svc/Send", errReader(0))
+	req.Header.Set("Accept", "application/protobuf")
+	req.Header.Set("Content-Type", "application/protobuf")
+	// Associate the cancellable context we just created to the request
+	req = req.WithContext(ctx)
+	// cancel context
+	time.Sleep(2 * time.Millisecond)
```
Comment (marioizquierdo) [inline internal/twirptest/proto/proto_test.go:89]: Instead of using time.Sleep, you should check the timeout channel like this:
```go
select {
case <-ctx.Done():
    // ctx.Err() will have context.DeadlineExceeded
case <-time.After(1 * time.Second):
    t.Fatalf("Waited too long for DeadlineExceeded ")
}
```

File: internal/twirptest/proto/proto_test.go [hunk 3]
```diff
@@ -13,8 +13,129 @@
...
+	}
+
+	expectedErrMessage := "context deadline exceeded"
+	if !strings.Contains(string(respBytes), expectedErrMessage) {
+		t.Errorf("twirp client err has unexpected message %q, want %q", string(respBytes), expectedErrMessage)
+	}
+}
+
+func TestReadRequestError(t *testing.T) {
+	// Test passes if this package compiles
```
Comment (marioizquierdo) [inline internal/twirptest/proto/proto_test.go:110]: Should remove this comment. For the test name, what about something more explicit like `TestFailedToReadRequestBodyError` ?

**Review by marioizquierdo — APPROVED** [review]
LGTM.

Let's release this as `v7.2.0`

File: protoc-gen-twirp/generator.go [hunk 4]
```diff
@@ -1334,6 +1334,13 @@ func (t *twirp) generateServerProtobufMethod(service *descriptor.ServiceDescript
 	t.P()
 	t.P(`  buf, err := `, t.pkgs["ioutil"], `.ReadAll(req.Body)`)
 	t.P(`  if err != nil {`)
+	t.P(`    if ctx.Err() == context.Canceled  {`)
+	t.P(`    		s.writeError(ctx, resp, twirp.NewError(twirp.Canceled, "context canceled"))`)
```
Comment (rhysh) [inline protoc-gen-twirp/generator.go:1338]: Optional: It seems unnecessary to write out the meaning of the error three times here (checking equality, setting Twirp status, adding message), and another three times in the next block. What do you think about replacing the last one of those with a call to `context.Canceled.Error()`, either via `ctx.Err().Error()` (since we have just determined that the two are `==`), or via `if ctxErr := context.Canceled; ctx.Err() == cxtErr /* ... */ ctxErr.Error()`?

It's not critical to change, and I'm not thrilled with any of the options I've provided for changing it. But it seems like more repetition than we need.

File: internal/twirptest/service_test.go [hunk 5]
```diff
@@ -1527,3 +1527,123 @@ func TestPanicFlushing(t *testing.T) {
...
+	}
+}
+
+func TestDeadlineExceededError(t *testing.T) {
+	// test case validate context deadline exceed error handling
+	var s Haberdasher
+	server := NewHaberdasherServer(s, nil)
+
+	ctx := context.Background()
+	ctx, _ = context.WithTimeout(ctx, 1*time.Millisecond)
```
Comment (rhysh) [inline internal/twirptest/service_test.go:1580]: Optional: When the ctx value is scoped to a function, WithTimeout and WithCancel usually get a `defer cancel()` right after to make that clear. For this, if it were my code, I'd give the CancelFunc a name (cancel) and immediately use it (as `defer cancel()`). Same for the previous test, though it already has a name there.

File: internal/twirptest/service_test.go [hunk 6]
```diff
@@ -1527,3 +1527,123 @@ func TestPanicFlushing(t *testing.T) {
 		t.Errorf("twirp client err has unexpected message %q, want %q", twerr.Msg(), "Internal service panic")
 	}
 }
+
+type errReader int
```
Comment (rhysh) [inline internal/twirptest/service_test.go:1531]: Optional: It looks like the value of this never gets used (but looks in some places like it _might_). I'd make this a `struct{}` so it's clear that it holds no useful information.

And an extra aside: I'd suggest also https://golang.org/pkg/testing/iotest/#ErrReader , but I'm surprised to see that it's new in Go 1.16, so is still too new to use in Twirp (we need to support at least the previous Go version, and maybe the one before that too).

**Review by rhysh — CHANGES_REQUESTED** [review]
The behavior change LGTM, but the tests are too far removed from how Twirp gets used in production. The tests should verify that we get the right behavior higher up in the stack, and demonstrate to readers (including our future selves) what that high-level behavior _is_.

File: internal/twirptest/service_test.go [hunk 7]
```diff
@@ -1527,3 +1527,123 @@ func TestPanicFlushing(t *testing.T) {
...
+		t.Fatalf("Waited too long for DeadlineExceeded ")
+	}
+	w := httptest.NewRecorder()
+	server.ServeHTTP(w, req)
+
+	// validate error code
+	expectedErrCode := twirp.ServerHTTPStatusFromErrorCode(twirp.DeadlineExceeded)
+	if w.Code != expectedErrCode {
+		t.Errorf("twirp ErrorCode expected to be %q, but found %q", expectedErrCode, w.Code)
+	}
```
[the remark this answers is not in the document]
Comment (chetanvalagit) [reply internal/twirptest/service_test.go:1622]: Checking as part of server hooks

File: protoc-gen-twirp/generator.go (hunk 4 above)
Comment (chetanvalagit) [reply protoc-gen-twirp/generator.go]: Done

File: internal/twirptest/service_test.go (hunk 5 above)
Comment (chetanvalagit) [reply internal/twirptest/service_test.go]: Agree. Done

File: internal/twirptest/service_test.go [hunk 8]
```diff
@@ -1527,3 +1527,123 @@ func TestPanicFlushing(t *testing.T) {
...
+	req = req.WithContext(ctx)
+
+	// wait for deadline exceed timeout in context object
+	select {
+	case <-ctx.Done():
+		// ctx.Err() will have context.DeadlineExceeded
+	case <-time.After(1 * time.Second):
+		t.Fatalf("Waited too long for DeadlineExceeded ")
+	}
+	w := httptest.NewRecorder()
```
[the remark this answers is not in the document]
Comment (chetanvalagit) [reply internal/twirptest/service_test.go:1614]: It difficult to generate error while reading body `buf, err := ioutil.ReadAll(req.Body)` and because of that it will be difficult to create production like behavior.

File: protoc-gen-twirp/generator.go [hunk 9]
```diff
@@ -1334,8 +1334,15 @@ func (t *twirp) generateServerProtobufMethod(service *descriptor.ServiceDescript
...
-	t.P(`    s.writeError(ctx, resp, wrapInternal(err, "failed to read request body"))`)
-	t.P(`    return`)
+	t.P(`  		if ctxErr := context.Canceled; ctxErr == ctx.Err() {`)
+	t.P(`    		s.writeError(ctx, resp, twirp.NewError(twirp.Canceled, ctxErr.Error()))`)
+	t.P(`    		return`)
+	t.P(`  		} else if ctxErr := context.DeadlineExceeded; ctxErr == ctx.Err() {`)
+	t.P(`    		s.writeError(ctx, resp, twirp.NewError(twirp.DeadlineExceeded, ctxErr.Error()))`)
+	t.P(`    		return`)
+	t.P(`  		}`)
+	t.P(`    	s.writeError(ctx, resp, twirp.NewError(twirp.Malformed, "failed to read request body"))`)
```
Comment (rhysh) [inline protoc-gen-twirp/generator.go:1344]: This case (returning the "malformed" code when the body read fails) is new since the first round. It's more of a behavior change but I think it's a good one, especially because the server returns that same status if the unmarshal step fails.

Comment (rhysh) [inline protoc-gen-twirp/generator.go:1344]: A couple lines down, this function uses the `malformedRequestError` helper to get a similar result in a different case. Please make both/all call sites look the same: this line should use the same helper, or all lines like that one should avoid it. (I don't know exactly why that helper exists.)

File: internal/twirptest/service_test.go [hunk 10]
```diff
@@ -1527,3 +1527,150 @@ func TestPanicFlushing(t *testing.T) {
...
+		t.Fatalf("Could not even read bytes from response: %q", err.Error())
+	}
+
+	expectedErrMessage := "context deadline exceeded"
+	if !strings.Contains(string(respBytes), expectedErrMessage) {
+		t.Errorf("twirp client err has unexpected message %q, want %q", string(respBytes), expectedErrMessage)
+	}
+}
+
+func TestFailedToReadRequestBodyError(t *testing.T) {
```
Comment (rhysh) [inline internal/twirptest/service_test.go:1636]: These three test cases have a lot of shared code. It's hard to see from a glance where they differ. Can you factor out the common behavior please?

It also gives a clear opportunity to add additional checks that should be true for all of these cases, including "because these are all client-initiated problems, the http status should be in the range [400,499]".

I find that a `testcase := func(differences) func(t *testing.T) {...}` at the start, followed by a bunch of calls to `t.Run("canceled-context", testcase(args))` is a helpful structure to share some behavior between tests.

File: internal/twirptest/service_test.go [hunk 11]
```diff
@@ -1527,3 +1527,150 @@ func TestPanicFlushing(t *testing.T) {
...
+		}
+		return ctx
+	}}
+	server := NewHaberdasherServer(s, hooks)
+	ctx := context.Background()
+
+	// Make a request, that will call endpoint method
+	req, _ := http.NewRequest(http.MethodPost, "http://testing:8080/twirp/twirp.internal.twirptest.Haberdasher/MakeHat", errReader{})
+	req.Header.Set("Accept", "application/protobuf")
+	req.Header.Set("Content-Type", "application/protobuf")
```
Comment (rhysh) [inline internal/twirptest/service_test.go:1654]: The tests should also cover behavior of the JSON client.

**Review by rhysh — CHANGES_REQUESTED** [review]

File: protoc-gen-twirp/generator.go (hunk 9 above)
Comment (chetanvalagit) [reply protoc-gen-twirp/generator.go:1344]: Yes, Original reported came back and ask me for this and shared thread. So thought let me cover that too which I missed at first place.

File: internal/twirptest/service_test.go (hunk 10 above)
Comment (chetanvalagit) [reply internal/twirptest/service_test.go]: I merged three test cases into one. 

We dont have any client and not making any client call. Instead calling server method directly as unit test case. Still wondering how I can check [400, 499] error?

Instead [400, 499] we are checking specific error https://github.com/twitchtv/twirp/pull/294/files#diff-f9d0e6645e9e1b67ccaaceb3da4d91900b84e0cf1ed5d2b7a94f85a9f0423870R1609. Wondering if that is enough?

File: internal/twirptest/service_test.go (hunk 11 above)
Comment (chetanvalagit) [reply internal/twirptest/service_test.go]: Thanks. I think error handling is missing for JSON client also and I end up adding that too. Please let me know if I am wrong.
https://github.com/twitchtv/twirp/pull/294/files#diff-ca331af0419d8cb15dfaf494e87bc6fac163b0f5a897b9927cd1eff3b422b2a8R1272-R1278

File: .github/workflows/pr-checks.yml [hunk 12]
```diff
@@ -29,6 +29,6 @@ jobs:
         import sys
         with open("cover_percentage", "r") as f:
           cov = float(f.readlines()[0])
-          if cov < 12.8:
+          if cov < 12.7:
```
Comment (chetanvalagit) [inline .github/workflows/pr-checks.yml:32]: I added code new code, added test case for it in one file (service_test.go). But I think not wise to add test case for each and every test case proto folders to make sure coverage % remains above 12.8.

File: internal/twirptest/service_test.go [hunk 13]
```diff
@@ -1527,3 +1527,127 @@ func TestPanicFlushing(t *testing.T) {
...
 	}
 }
+
+type errReader struct{}
+
+func (errReader) Read(p []byte) (n int, err error) {
+	return 0, fmt.Errorf("test error")
+}
+
+func TestRequestBodyError(t *testing.T) {
```
Comment (marioizquierdo) [inline internal/twirptest/service_test.go:1537]: Nice update on the tests!

If you can, it would be nice to also add a comment on top of the text to explain a little bit what is trying to accomplish. For example:
```go
// Test failures when reading the request.
// The request could be incomplete because the client timed out, which in most cases is signaled by
// a context cancelation or timeout. In those cases, the twirp handled would properly identify and return those errors.
func TestRequestBodyError(t *testing.T) {
  // ...
}
```

**Review by marioizquierdo — CHANGES_REQUESTED** [review]

File: protoc-gen-twirp/generator.go [hunk 14]
```diff
@@ -1146,6 +1146,20 @@ func (t *twirp) generateServer(file *descriptor.FileDescriptorProto, service *de
...
+	t.P(`func (s *`, servStruct, `) handleRequestBodyError(ctx `, t.pkgs["context"], `.Context, resp `, t.pkgs["http"], `.ResponseWriter, msg string) {`)
+	t.P(`  if ctxErr := context.Canceled; ctxErr == ctx.Err() {`)
+	t.P(`    s.writeError(ctx, resp, twirp.NewError(twirp.Canceled, ctxErr.Error()))`)
+	t.P(`    return`)
+	t.P(`  } else if ctxErr := context.DeadlineExceeded; ctxErr == ctx.Err() {`)
+	t.P(`    s.writeError(ctx, resp, twirp.NewError(twirp.DeadlineExceeded, ctxErr.Error()))`)
+	t.P(`    return`)
+	t.P(`  }`)
+	t.P(`  s.writeError(ctx, resp, malformedRequestError(msg))`)
+	t.P(`}`)
```
[the remark this answers is not in the document]
Comment (chetanvalagit) [reply protoc-gen-twirp/generator.go:1161]: Done

File: protoc-gen-twirp/generator.go [hunk 15]
```diff
@@ -1146,6 +1146,20 @@ func (t *twirp) generateServer(file *descriptor.FileDescriptorProto, service *de
...
 
+	// Write request body read issue
+	t.P(`// handleRequestBodyError is used to handle error when the twirp server cannot read request`)
+	t.P(`func (s *`, servStruct, `) handleRequestBodyError(ctx `, t.pkgs["context"], `.Context, resp `, t.pkgs["http"], `.ResponseWriter, msg string) {`)
+	t.P(`  if ctxErr := context.Canceled; ctxErr == ctx.Err() {`)
+	t.P(`    s.writeError(ctx, resp, twirp.NewError(twirp.Canceled, ctxErr.Error()))`)
+	t.P(`    return`)
+	t.P(`  } else if ctxErr := context.DeadlineExceeded; ctxErr == ctx.Err() {`)
+	t.P(`    s.writeError(ctx, resp, twirp.NewError(twirp.DeadlineExceeded, ctxErr.Error()))`)
+	t.P(`    return`)
```
[the remark this answers is not in the document]
Comment (chetanvalagit) [reply protoc-gen-twirp/generator.go:1158]: Done

File: internal/twirptest/service_test.go (hunk 13 above)
Comment (chetanvalagit) [reply internal/twirptest/service_test.go:1540]: Done

File: internal/twirptest/service_test.go [hunk 16]
```diff
@@ -1527,3 +1527,131 @@ func TestPanicFlushing(t *testing.T) {
...
+			}
+
+			// validate error message
+			respBytes, err := ioutil.ReadAll(w.Body)
+			if err != nil {
+				t.Fatalf("Could not even read bytes from response: %q", err.Error())
+			}
+
+			expectedErrMessage := tc.errorString
+			fmt.Println(string(respBytes))
```
Comment (marioizquierdo) [inline internal/twirptest/service_test.go:1651]: Println running wild!

Comment (chetanvalagit) [reply internal/twirptest/service_test.go:1651]: Ops good catch. Added for testing.

File: protoc-gen-twirp/generator.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (marioizquierdo) [reply protoc-gen-twirp/generator.go:1158]: To clarify, I also find the `ctxErr` assignment inside the conditional un-necessarily confusing to read, and it obfuscates the final error message. A developer reading this code may be trying to figure out why they see "context canceled" in the logs, it would be better IMO if we would make that as explicit as possible. `context.Canceled.Error()` is more explicit than `ctxErr.Error()`, specially if that same line means something different on each conditional.

I would prefer if we write this in a more explicit form like this:

```go
func (s *svc2Server) handleRequestBodyError(ctx context.Context, resp http.ResponseWriter, err error, msg string) {
    if ctx.Err() == context.Canceled {
        s.writeError(ctx, resp, twirp.NewError(twirp.Canceled, "failed to read request: context canceled"))
        return
    }
    if ctx.Err() == context.DeadlineExceeded {
        s.writeError(ctx, resp, twirp.NewError(twirp.DeadlineExceeded, "failed to read request: deadline exceeded"))
        return
    }
    s.writeError(ctx, resp, twirp.WrapError(malformedRequestError(msg), err)) // original error can be unwrapped for inspection
}
```
The reason is that this is both easier to read and easier to find when searching for the error message.

I know @chetanvalagit has been going back and forth with this and may not want to keep running in circles, this is more of a question for @rhysh, and while I have a preference here, I don't think this should block the PR any longer, so I'm fine with whatever you both decide now.

File: protoc-gen-twirp/generator.go [hunk 17]
```diff
@@ -1269,8 +1284,8 @@ func (t *twirp) generateServerJSONMethod(service *descriptor.ServiceDescriptorPr
 	t.P(`  reqContent := new(`, t.goTypeName(method.GetInputType()), `)`)
 	t.P(`  unmarshaler := `, t.pkgs["jsonpb"], `.Unmarshaler{AllowUnknownFields: true}`)
 	t.P(`  if err = unmarshaler.Unmarshal(req.Body, reqContent); err != nil {`)
-	t.P(`    s.writeError(ctx, resp, malformedRequestError("the json request could not be decoded"))`)
-	t.P(`    return`)
+	t.P(`       s.handleRequestBodyError(ctx, resp, "the json request could not be decoded", err)`)
+	t.P(`    	return`)
```
Comment (marioizquierdo) [inline protoc-gen-twirp/generator.go:1288]: This still has inconsistent indentation. Notice the first line has 3 spaces and the second has 1 tab. The indentation here is maintained manually because it is inside strings, and then is automatically corrected after the code is generated, but it is still useful to indent properly for readability

Comment (chetanvalagit) [reply protoc-gen-twirp/generator.go]: Fixed

File: protoc-gen-twirp/generator.go [hunk 18]
```diff
@@ -1146,6 +1146,21 @@ func (t *twirp) generateServer(file *descriptor.FileDescriptorProto, service *de
 	t.P(`}`)
 	t.P()
 
+	// Write request body read issue
+	t.P(`// handleRequestBodyError is used to handle error when the twirp server cannot read request`)
+	t.P(`func (s *`, servStruct, `) handleRequestBodyError(ctx `, t.pkgs["context"], `.Context, resp `, t.pkgs["http"], `.ResponseWriter, msg string, err error) {`)
+	t.P(`  if ctxErr := context.Canceled; ctxErr == ctx.Err() {`)
+	t.P(`    s.writeError(ctx, resp, twirp.NewError(twirp.Canceled, ctxErr.Error()))`)
```
Comment (rhysh) [inline protoc-gen-twirp/generator.go:1153]: It looks like all references to imported packages need to go via the `t.pkgs` map, as they do on the line above. That would hold for the `context` and `twirp` references here and in the next stanza, and for the following `twirp` reference.

(It's unfortunate that we don't have tests to detect that; I don't think it could break based on user input (the name of the protobuf package, for example) since "our" calls to `registerPackageName` come so early, so I'd be surprised if this would be the only part of the generator that skipped the map.)

File: internal/twirptest/service_test.go [hunk 19]
```diff
@@ -1527,3 +1527,130 @@ func TestPanicFlushing(t *testing.T) {
...
+			} else {
+				defer cancel()
+			}
+			w := httptest.NewRecorder()
+			server.ServeHTTP(w, req)
+
+			// validate error code
+			expectedErrCode := twirp.ServerHTTPStatusFromErrorCode(tc.expectedError)
+			if w.Code != expectedErrCode {
+				t.Errorf("twirp ErrorCode expected to be %q, but found %q", expectedErrCode, w.Code)
```
Comment (rhysh) [inline internal/twirptest/service_test.go:1641]: The Errorf text isn't right: this compares HTTP status codes.

File: internal/twirptest/service_test.go [hunk 20]
```diff
@@ -1527,3 +1527,130 @@ func TestPanicFlushing(t *testing.T) {
...
+// a context cancelation or timeout. In those cases, the twirp handled would properly identify and return those errors.
+func TestRequestBodyError(t *testing.T) {
+	type contextUpdate func(ctx context.Context) (context.Context, context.CancelFunc)
+	testCases := []struct {
+		testname         string
+		expectedError    twirp.ErrorCode
+		errorString      string
+		reqestType       string
+		contextFunc      contextUpdate
+		isContextCancled bool
```
Comment (rhysh) [inline internal/twirptest/service_test.go:1548]: If you're making other changes, please fix the spelling here: `isContextCanceled`.

File: internal/twirptest/service_test.go [hunk 21]
```diff
@@ -1527,3 +1527,130 @@ func TestPanicFlushing(t *testing.T) {
...
+// Test failures when reading the request.
+// The request could be incomplete because the client timed out, which in most cases is signaled by
+// a context cancelation or timeout. In those cases, the twirp handled would properly identify and return those errors.
+func TestRequestBodyError(t *testing.T) {
+	type contextUpdate func(ctx context.Context) (context.Context, context.CancelFunc)
+	testCases := []struct {
+		testname         string
+		expectedError    twirp.ErrorCode
+		errorString      string
+		reqestType       string
```
Comment (rhysh) [inline internal/twirptest/service_test.go:1546]: If you're making other changes, please fix the spelling here: `requestType`.

File: protoc-gen-twirp/generator.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (rhysh) [reply protoc-gen-twirp/generator.go:1158]: Either is fine, provided the spelling is right. Pulling this out into a specialized function helps with safety, and the tests help with safety. Those two alone are probably sufficient protection against 1/ "cancelled" vs "canceled" or 2/ copy–paste errors that turn "context canceled" into "deadline exceeded".

**Review by rhysh — APPROVED** [review]
Thanks for restructuring and expanding the test code. I don't think that any of my remaining comments are deal-breakers. They'd be good to change, but don't need another round of review.

File: protoc-gen-twirp/generator.go (hunk 18 above)
Comment (marioizquierdo) [reply protoc-gen-twirp/generator.go:1153]: Nice catch! and good point about the tests, thanks Rhysh

Comment (chetanvalagit) [reply protoc-gen-twirp/generator.go]: Great catch :)
Fixed.

**Review by marioizquierdo — APPROVED** [review]
LGTM
